### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

In [2]:
# Calculate unique players using value counts to remove duplicates
player_unique = purchase_data["SN"].unique()
player_count = len(player_unique)
display =pd.DataFrame({"Number of Unique Players": [player_count]})
display

Number of Unique Players
0                       576

## Purchasing Analysis (Total)

In [18]:
# Define Variables
purchase_count = purchase_data["SN"].count()
item_unique = purchase_data["Item ID"].unique()
item_count = len(item_unique)
total_revenue = sum(purchase_data["Price"])
avg_price = total_revenue / purchase_count 

# Create summary table
summary_df =pd.DataFrame({"Number of Unique Items": [item_count], "Average Price":  [avg_price],
                          "Number of Purchases":  [purchase_count], "Total Revenue": [total_revenue]})

# Map format summary table
summary_df["Average Price"] = summary_df["Average Price"].map("${:.2f}".format)
summary_df["Total Revenue"] = summary_df["Total Revenue"].map("${:.2f}".format)

summary_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780      $2379.77

## Gender Demographics

In [16]:
# Create Unique Players Dataframe removing duplicates 
player_unique_df = purchase_data.drop_duplicates('SN', keep = 'first') 
player_unique_df["Gender"].value_counts()

# Percentage and Count of Male Players
only_male = player_unique_df.loc[player_unique_df["Gender"] == "Male",]
male_percent = len(only_male) / player_count 

# Percentage and Count of Female Players
only_female = player_unique_df.loc[player_unique_df["Gender"] == "Female",]
female_percent = len(only_female) / player_count

# Percentage and Count of Other / Non-Disclosed
only_other = player_unique_df.loc[player_unique_df["Gender"] == "Other / Non-Disclosed",]
other_percent = (len(only_other) / player_count) 

# Create summary table 
summary2 = pd.DataFrame(
    {"Gender": ["Male", "Female", "Other"],
     "Total Count": [len(only_male), len(only_female), len(only_other)],
     "Percentage of Players": [male_percent, female_percent, other_percent]})

# Style format summary table
summary2.style.format({'Percentage of Players': "{:.2%}"})


## Purchasing Analysis (Gender)

In [24]:
#Calculate total purchases by gender
male_purchases = purchase_data.loc[purchase_data["Gender"] == "Male",]
female_purchases = purchase_data.loc[purchase_data["Gender"] == "Female",]
other_purchases = purchase_data.loc[purchase_data["Gender"] == "Other / Non-Disclosed",]

#Calculate total Revenue by gender
male_revenue = sum(male_purchases["Price"])
female_revenue = sum(female_purchases["Price"])
other_revenue = sum(other_purchases["Price"])

# Create Summary Table 
summary3_df  = pd.DataFrame({"Gender": ["Male", "Female", "Other", "Total" ],
     "Purchase Count": [len(male_purchases), len(female_purchases), len(other_purchases), purchase_count],
    "Average Purchase Price": [male_revenue/len(male_purchases), female_revenue/len(female_purchases), 
                                     other_revenue/len(other_purchases), avg_price],
          "Total Purchase Value": [male_revenue, female_revenue, other_revenue, total_revenue],
     "Avg Total Purchase per Person": [male_revenue/len(only_male), female_revenue/len(only_female), 
                                     other_revenue/len(only_other), total_revenue/player_count]})

# Formatting 
summary3_df["Average Purchase Price"] = summary3_df["Average Purchase Price"].map("${:.2f}".format)
summary3_df["Total Purchase Value"] = summary3_df["Total Purchase Value"].map("${:.2f}".format)
summary3_df["Avg Total Purchase per Person"] = summary3_df["Avg Total Purchase per Person"].map("${:.2f}".format)
summary3_df

Gender  Purchase Count Average Purchase Price Total Purchase Value  \
0    Male             652                  $3.02             $1967.64   
1  Female             113                  $3.20              $361.94   
2   Other              15                  $3.35               $50.19   
3   Total             780                  $3.05             $2379.77   

  Avg Total Purchase per Person  
0                         $4.07  
1                         $4.47  
2                         $4.56  
3                         $4.13

## Age Demographics

In [25]:
# Create bins in which to place values based upon TED Talk views
bins = [0, 9.9, 14.9, 19.9, 24.9, 29.9,
        34.9, 39.9, 45]

# Create labels for these bins
group_labels = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34",
                "35-39", "40+"]

# Slice the data and place it into bins
pd.cut(player_unique_df["Age"], bins, labels=group_labels).head()

0    20-24
1      40+
2    20-24
3    20-24
4    20-24
Name: Age, dtype: category
Categories (8, object): [<10 < 10-14 < 15-19 < 20-24 < 25-29 < 30-34 < 35-39 < 40+]

In [34]:
# Place the data series into a new column inside of the DataFrame
player_unique_df["Age Group"] = pd.cut(player_unique_df["Age"], bins, labels=group_labels)

# Create a GroupBy object based upon "Age Group"
player_unique_df_group = player_unique_df.groupby("Age Group")

# Get the average of each bin within the GroupBy object
summary4_df = player_unique_df_group[["SN"]].agg(['count'])

#Add Percentage Column 
summary4_df["Percentage of Players"] = player_unique_df_group["SN"].count()/len(player_unique_df)*100

#Format and Display Table 
summary4_df["Percentage of Players"] = summary4_df["Percentage of Players"].map("{:.2f}%".format)
summary4_df

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


SN Percentage of Players
          count                      
Age Group                            
<10          17                 2.95%
10-14        22                 3.82%
15-19       107                18.58%
20-24       258                44.79%
25-29        77                13.37%
30-34        52                 9.03%
35-39        31                 5.38%
40+          12                 2.08%

## Purchasing Analysis (Age)

In [37]:
# Slice the data and place it into bins
pd.cut(purchase_data["Age"], bins, labels=group_labels).head()

# Place the data series into a new column inside of the DataFrame
purchase_data["Age Group"] = pd.cut(purchase_data["Age"], bins, labels=group_labels)
purchase_data.head()

# Create a GroupBy object based upon "Age Group"
purchase_data_group = purchase_data.groupby("Age Group")

# Get the average of each column within the GroupBy object
summary5_df = purchase_data_group[["Price"]].agg(['count', 'mean', 'sum'])

#Add Column and rename headings 
summary5_df["Avg Purchase per Person"] = purchase_data_group["Price"].sum()/ player_unique_df_group["SN"].count()
summary5_df = summary5_df.rename(columns={"count": "Purchase Count", "mean": "Average Purchase Price", "sum": "Total Purchase Value" })

#Format and display
summary5_df["Avg Purchase per Person"] = summary5_df["Avg Purchase per Person"].map("${:.2f}".format)
summary5_df


Price                                              \
          Purchase Count Average Purchase Price Total Purchase Value   
Age Group                                                              
<10                   23               3.353478                77.13   
10-14                 28               2.956429                82.78   
15-19                136               3.035956               412.89   
20-24                365               3.052219              1114.06   
25-29                101               2.900990               293.00   
30-34                 73               2.931507               214.00   
35-39                 41               3.601707               147.67   
40+                   13               2.941538                38.24   

          Avg Purchase per Person  
                                   
Age Group                          
<10                         $4.54  
10-14                       $3.76  
15-19                       $3.86  
20-24                       $4.32  
25-29                       $3.81  
30-34                       $4.12  
35-39                       $4.76  
40+                         $3.19

## Top Spenders

In [39]:
# Calculate Purchase Count
purchase_data_group2 = purchase_data.groupby(['SN'])
b = pd.DataFrame(purchase_data_group2["Purchase ID"].count())

#Calculate Average Purchase Price and Total Purchase Value 
b["Average Purchase Price"] = pd.DataFrame(purchase_data_group2["Price"].mean())
b["Total Purchase Value"] = b["Average Purchase Price"] * b["Purchase ID"]

#Sort the total purchase value column in descending order
c = b.sort_values("Total Purchase Value", ascending=False)

# Format, rename and display
c["Total Purchase Value"] = c["Total Purchase Value"].map("${:.2f}".format)
c["Average Purchase Price"] = c["Average Purchase Price"].map("${:.2f}".format)
c = c.rename(columns={"Purchase ID": "Purchase Count"})
c.head()

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

In [43]:
#Retrieve the Item ID, Item Name, and Item Price columns
d = purchase_data[["Item ID", "Item Name", "Price"]]

#Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
d = d.groupby(['Item ID', "Item Name"])
e = pd.DataFrame(d["Item Name"].count())
e["Item Price"] = d["Price"].mean()
e["Total Purchase Value"] = e["Item Name"] * e["Item Price"]
e = e.rename(columns={"Item Name": "Purchase Count"})

#Sort the purchase count column in descending order
f = e.sort_values("Purchase Count", ascending=False)

# Format and display
f["Total Purchase Value"] = f["Total Purchase Value"].map("${:.2f}".format)
f["Item Price"] = f["Item Price"].map("${:.2f}".format)
f.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [44]:
#Sort the above table by total purchase value in descending order
g = e.sort_values("Total Purchase Value", ascending=False)

#Format & Display
g["Total Purchase Value"] = g["Total Purchase Value"].map("${:.2f}".format)
g["Item Price"] = g["Item Price"].map("${:.2f}".format)
g.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## # Three Observable Trends 

1. The large majority of HoP players are male!
Perhaps it does not come as a surprise given the overall gender demographic of the gaming community and given the nature of the fantasy game genre, that the large majority of players are male. It would be interesting to explore this against data on number of female gamers across the industry and in this specific sub-genre. In addition, an analysis of how this trend has changed over time would be perhaps helpful in determining how the game may be promoted and  advertised in future. 


2. The main demographic for purchases is age group 20-24
Both in terms of the player numbers and revenues earned, the 20-24 age group forms the single biggest segment representing over 40% of the totals. When combined with the two closest age groups, the 15-30 year age demographic forms over 75% of the numbers. While this trend by itself may not be very insightful as it perhaps does stand out from the industry norm, when combined with other broader data sources (eg payment methods used, social media feeds, or frequency/timing of purchases) it can allow the company to paint a better picture of the behavior trends for this key demographic. 


3. The purchases are broadly distributed in terms of popularity and revenues.
While the data clearly identifies "Oathbreaker, , Last Hope of the Breaking Storm" as the most popular (12 purchases) and highest revenue earning Item($50.76), there are a host of other items which are close following seconds with 7-9 purchases and $35-45 in revenues earned. As such this does not point to a single item or area where the company can focus on to increase purchases or revenues but nevertheless a broad distribution points to stability in revenues in case one of the key items loses popularity for any reason. Perhaps, further breakdown of the data into various item categories or game segments will be able to identify further trends which can be exploited and explored.     


